### DISCUSSION TITLE
[Feature] Reducing dimension in labeled metric value query

### DISCUSSION THREAD

**Discussion by weixiang1862** [apache/skywalking#11031] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

I was working on mongodb-monitoring recently:
|MongoDB|skywalking|
|-----------|------------|
|cluster|service|
|Node|Instance|
|ReplSet|Endpoint|

The targets in otel-collector is every node in cluster:
```
scrape_configs:
    - job_name: 'mongodb-monitoring'
      static_configs:
      - targets: ['localhost:19216',''localhost:29216'',''localhost:39216'']
         labels:
             cluster: test-cluster
```
The /metric endpoint provide connection metrics:
```
mongodb_ss_connections{conn_type="current"} 10
```
I want to monitoring `connection` in cluster dimension, the mal is defined like this:
```
- name: connections
   exp: mongodb_ss_connections.tagEqual('conn_type', 'current').sum(['cluster'])
```
I excepted cluster connection metric aggregated like this:
```
mongodb_ss_connections{conn_type="current", service_instance_id="localhost:19216"} 10
mongodb_ss_connections{conn_type="current", service_instance_id="localhost:29216"} 10   --sum(['cluster'])-->  30  --avg--> 30
mongodb_ss_connections{conn_type="current", service_instance_id="localhost:39216"} 10
```
But the rea like this, because mal aggregate metric by every target: 
```
mongodb_ss_connections{conn_type="current", service_instance_id="localhost:19216"} 10   --sum(['cluster'])-->  10
mongodb_ss_connections{conn_type="current", service_instance_id="localhost:29216"} 10   --sum(['cluster'])-->  10   --avg--> 10
mongodb_ss_connections{conn_type="current", service_instance_id="localhost:39216"} 10   --sum(['cluster'])-->  10
```
So I change the mal to this, store every node connections in labeld value:
```
- name: connections
   exp: mongodb_ss_connections.tagEqual('conn_type', 'current').sum(['cluster','service_instance_id'])
```
Now I have connection sum of every node in labelvalue mode:
```
localhost:19216,10|localhost:29216,10|localhost:39216,10
```
I want the labeled value cant be query like this:
```
localhost:19216,10|localhost:29216,10|localhost:39216,10  --> 30
```
There are two place we could modify to support above:
**1.MQE**
```
# reducingDimension is a function cal labeldvalue in sum alg and turned labeledvalue to value
reducingDimension(mongodb_cluster_connections, SUM)
```
**2.graphQL readLabeledMetricsValues**
```
# add a parameter ReducingDimensionAlg in readLabeledMetricsValues method
readLabeledMetricsValues(MetricsCondition condition,List<String> labels,ReducingDimensionAlg alg,Duration duration) 
```

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
If you want to manipulate labels, you could use closure to process it first.
There is no automatical way like you mentioned.

**Comment by wu-sheng** [comment]
You even don't have cluster label, how could it possible it works like that?

**Reply by weixiang1862** [reply]
cluster label was added manually by otel promethues exporter.
sum(['cluster']) get value 10 but not 30(3 node actually total connection num is 30).
Does result 10 meet expectations? or any problems with cluster level mal?

**Comment by weixiang1862** [comment]
I think it is default to summation every node connections in current mechanism. 
From debugging result, I see otel-collector can not guarantee report every node's metrics in one grpc call, and skywalking also not aggregate them in one round. It cause the sum(['cluster']) result inconsistent with the facts.

**Comment by wu-sheng** [comment]
You need sum for a bulk, and sum downsampling.

**Reply by weixiang1862** [reply]
Use sum downsampling means hour and day metric value would grow terrifying bigger，it can not reflect the real level of cluster connection number at hour or day view.

**Reply by wu-sheng** [reply]
You have to choose. Otherwise, this is a logic issue rather than code issue. Even human can't determine that.

**Reply by weixiang1862** [reply]
Yes, you are totally right. It is hard to get a balance between accuracy and code implmention. And that also the reason why I request you import a new function in MQE. This function is designed for this scene. I may not describe it clearly in above.So let me   try to describe it clearly and briefly again:

1. sum(['cluster', 'service_instance_id']) can aggregate metrics accurately for every node connections, even though nodes metrics not reported in a grpc call. The avglabeled metrics finally like this:
```
node1,10|node2,10|node3,10
```
2. the next step I need to do is calculate the sum of three node，the function for MQE I mentioned above do this calculation job，the calculation algorithm not just sum, it also can be avg max min... This function turned a labeledvalue to a singlevalue
3. the dashboard can show the calculation result 30，it represents the real value of the whole cluster in minutes，hour and day scale.

This function is not just use in mongodb monitoring，I found the same problem in redis and other middleware monitoring.

Thank you for listening me this long paragraph!

**Reply by wu-sheng** [reply]
@wankai123 What is the status of aggregating labels of a metric in the MQE?

**Reply by wankai123** [reply]
> @wankai123 What is the status of aggregating labels of a metric in the MQE?

no，MQE has no such func to aggr all labels value in one value yet

**Reply by wu-sheng** [reply]
@weixiang1862 Then, you could propose a function at the MQE level. Kai and I could review that.

**Reply by weixiang1862** [reply]
ok，I will do some works in the next few days.
